File: Javascript/15. 3Sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  nums.sort((a, b) => a - b);
  let res = [];

  for (let i = nums.length - 1; i >= 2; i--) {
    if (nums[i] === nums[i + 1]) continue;

    let left = 0;
    let right = i - 1;

    while (left < right) {
      const [a, b, c] = [nums[left], nums[right], nums[i]];

      const total = a + b + c;

      if (total === 0) {
        res.push([a, b, c]);
        left++;
        right--;

        while (left < right && nums[left] === nums[left - 1]) left++;
        while (left < right && nums[right] === nums[right + 1]) right--;
      } else {
        if (total < 0) left++;
        else right--;
      }
    }
  }

  return res;
};

File: Python/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        n = len(nums)
        visited = set()

        for i in range(n):
            low = i + 1
            high = n - 1
            curr = nums[i]
            if (curr in visited):
                continue
            visited.add(curr)
            subvisited = set()
            while (low < high):
                l = nums[low]
                r = nums[high]
                delta = curr + r + l
                if (delta > 0):
                    high -= 1
                elif (delta < 0):
                    low += 1
                else:
                    if (l not in subvisited):
                        res.append([curr, l, r])
                    subvisited.add(l)
                    low += 1
        return res
